libafl/src/inputs/encoded.rs

```diff
@@ -72,14 +96,76 @@ where
 }
 
 impl InputDecoder for TokenInputEncoderDecoder {
-    fn decode(&self, input: &EncodedInput, bytes: &mut Vec<u8>) -> Result<(), Error> {
-        for id in input.codes() {
+    fn decode(&mut self, input: &mut EncodedInput, bytes: &mut Vec<u8>) -> Result<(), Error> {
+        let codes = &mut input.codes;
+        if self.encoding_type == TokenizationKind::WithWhitespace {
+            // If we encode whitespace, mutations can f*ck up validity
+            // by removing whitespace or single char tokens like `+` to
+            // create illegal entries like Variable1"String". We need to
+            // fix such entries. So wherever we have two tokens that have
+            // a length larger than 1 and are not whitespace, we randomly
+            // insert whitespace or a single byte token from our token vec.
+            // We do this randomly because this enhances the overall
+            // mutation quality.
+            let mut positions: Vec<usize> = vec![];
+            let mut prev_len = 0;
+
+            for pos in 0..codes.len() {
+                let tok = self
+                    .id_table
+                    .get(&(codes[pos] % self.next_id))
+                    .ok_or_else(|| {
+                        Error::illegal_state(format!("Id {} not in the decoder table", codes[pos]))
+                    })?;
+                let len = tok.len();
+                if prev_len > 1 && len > 1 {
+                    positions.push(pos);
+                }
+                prev_len = len;
+            }
+
+            if positions.len() > 0 {
+                // We need to repair the input
+                let mut rand = StdRand::new();
+
+                // we have to walk backwards otherwise the positions become wrong
+                for pos in positions.len() - 1..0 {
```

Comment:
This doesn't actually count backwards, I think, you want to use `.rev()`
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.rev

Reply:
Something like
`for (pos, val) in positions.iter().enumerate().rev() {`